File: core/src/main/java/javabot/web/resources/PublicOAuthResource.java
package javabot.web.resources;

import com.antwerkz.sofia.Sofia;
import com.codahale.metrics.annotation.Timed;
import com.google.common.base.Optional;
import javabot.dao.AdminDao;
import javabot.model.Admin;
import javabot.web.JavabotConfiguration;
import javabot.web.model.Authority;
import javabot.web.model.InMemoryUserCache;
import javabot.web.model.OAuthConfig;
import javabot.web.model.User;
import org.brickred.socialauth.AuthProvider;
import org.brickred.socialauth.Profile;
import org.brickred.socialauth.SocialAuthConfig;
import org.brickred.socialauth.SocialAuthManager;
import org.brickred.socialauth.util.SocialAuthUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.NewCookie;
import javax.ws.rs.core.Response;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Path("/auth")
@Produces(MediaType.TEXT_HTML)
public class PublicOAuthResource {

    @Inject
    private AdminDao adminDao;

    private static final Logger log = LoggerFactory.getLogger(PublicOAuthResource.class);
    public static final String AUTH_MANAGER = "authManager";

    private JavabotConfiguration configuration;

    @GET
    @Path("/login")
    public Response requestOAuth(@Context HttpServletRequest request) throws URISyntaxException {
        List<OAuthConfig> oauthCfg = configuration.getOAuthCfg();
        if (oauthCfg != null) {
            try {
                SocialAuthManager manager = getSocialAuthManager();

                request.getSession().setAttribute(AUTH_MANAGER, manager);

                final URI uri = new URI(manager.getAuthenticationUrl("googleplus", configuration.getOAuthSuccessUrl()));
                return Response
                           .temporaryRedirect(uri)
                           .build();
            } catch (Exception e) {
                log.error(e.getMessage(), e);
            }
        }
        throw new WebApplicationException(Response.Status.BAD_REQUEST);
    }

    /**
     * Handles the OAuth server response to the earlier AuthRequest
     *
     * @return The OAuth identifier for this user if verification was successful
     */
    @GET
    @Timed
    @Path("/verify")
    public Response verifyOAuthServerResponse(@Context HttpServletRequest request) {
        SocialAuthManager manager = (SocialAuthManager) request.getSession().getAttribute(AUTH_MANAGER);

        if (manager != null) {
            try {
                Map<String, String> params = SocialAuthUtil.getRequestParametersMap(request);
                AuthProvider provider = manager.connect(params);

                Profile p = provider.getUserProfile();

                Sofia.loggingInUser(p);

                User tempUser = new User(UUID.randomUUID());
                tempUser.setOpenIDIdentifier(p.getValidatedId());
                tempUser.setOAuthInfo(provider.getAccessGrant());

                tempUser.setEmail(p.getEmail());

                tempUser.getAuthorities().add(Authority.ROLE_PUBLIC);

                Optional<User> userOptional = InMemoryUserCache.INSTANCE.getByOpenIDIdentifier(tempUser.getOpenIDIdentifier());
                if (!userOptional.isPresent()) {
                    Admin admin = adminDao.getAdminByEmailAddress(tempUser.getEmail());
                    if(admin != null) {
                        tempUser.getAuthorities().add(Authority.ROLE_ADMIN);
                    }
                    InMemoryUserCache.INSTANCE.put(tempUser);
                } else {
                    tempUser = userOptional.get();
                }

                return Response.temporaryRedirect(new URI("/"))
                               .cookie(replaceSessionTokenCookie(Optional.of(tempUser)))
                               .build();
            } catch (Exception e) {
                log.error(e.getMessage(), e);
            }
        }

        // Must have failed to be here
        throw new WebApplicationException(Response.Status.UNAUTHORIZED);
    }

    /**
     * Gets an initialized SocialAuthManager
     *
     * @return gets an initialized SocialAuthManager
     */
    private SocialAuthManager getSocialAuthManager() {
        SocialAuthConfig config = SocialAuthConfig.getDefault();
        try {
            config.load(configuration.getOAuthCfgProperties());
            SocialAuthManager manager = new SocialAuthManager();
            manager.setSocialAuthConfig(config);
            return manager;
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        return null;
    }

    protected NewCookie replaceSessionTokenCookie(Optional<User> user) {
        if (user.isPresent()) {
            String value = user.get().getSessionToken().toString();
            log.debug("Replacing session token with {}", value);
            return new NewCookie(JavabotConfiguration.SESSION_TOKEN_NAME, value, "/", null, null, 86400 * 30, false);
        } else {
            // Remove the session token cookie
            log.debug("Removing session token");
            return new NewCookie(JavabotConfiguration.SESSION_TOKEN_NAME, null, null, null, null, 0, false);
        }
    }

    public void setConfiguration(final JavabotConfiguration configuration) {
        this.configuration = configuration;
    }

    public JavabotConfiguration getConfiguration() {
        return configuration;
    }
}

File: src/main/kotlin/javabot/web/resources/PublicOAuthResource.kt
package javabot.web.resources

import com.google.common.base.Optional
import com.google.inject.Injector
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.servlet.http.HttpServletRequest
import jakarta.ws.rs.GET
import jakarta.ws.rs.Path
import jakarta.ws.rs.Produces
import jakarta.ws.rs.WebApplicationException
import jakarta.ws.rs.core.Context
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.NewCookie
import jakarta.ws.rs.core.Response
import jakarta.ws.rs.core.Response.Status.BAD_REQUEST
import jakarta.ws.rs.core.Response.Status.UNAUTHORIZED
import java.net.URI
import java.net.URISyntaxException
import javabot.dao.AdminDao
import javabot.model.Admin
import javabot.web.JavabotConfiguration
import javabot.web.model.Authority.ROLE_ADMIN
import javabot.web.model.Authority.ROLE_PUBLIC
import javabot.web.model.InMemoryUserCache.INSTANCE
import javabot.web.model.User
import org.eclipse.microprofile.config.inject.ConfigProperty
import org.slf4j.LoggerFactory

@Path("/auth")
@Produces(MediaType.TEXT_HTML)
@ApplicationScoped
class PublicOAuthResource @Inject constructor(private val injector: Injector) {

    // AdminDao is Guice-managed, not a CDI bean -- see GuiceInjectorProducer.
    private val adminDao: AdminDao by lazy { injector.getInstance(AdminDao::class.java) }

    @ConfigProperty(name = "javabot.oauth.success.url", defaultValue = "/")
    lateinit var oauthSuccessUrl: String

    // Optional<String> rather than a nullable String: MicroProfile Config treats an unset
    // property on a plain String field as required even with an empty defaultValue.
    @ConfigProperty(name = "javabot.oauth.config")
    lateinit var oauthConfigPath: java.util.Optional<String>

    @GET
    @Path("/login")
    @Throws(URISyntaxException::class)
    fun requestOAuth(@Context request: HttpServletRequest): Response {
        if (oauthConfigPath.isPresent && oauthConfigPath.get().isNotEmpty()) {
            // If the user already has a session cookie, they're considered authenticated.
            val user =
                INSTANCE.getBySessionToken(
                    request.cookies
                        ?.firstOrNull { it.name == JavabotConfiguration.SESSION_TOKEN_NAME }
                        ?.value
                )
            if (user != null) {
                user.authorities.add(ROLE_PUBLIC)
                val admin = adminDao.getAdminByEmailAddress(user.email)
                if (admin != null) {
                    user.authorities.add(ROLE_ADMIN)
                }
                INSTANCE.put(user)
                return Response.temporaryRedirect(URI(oauthSuccessUrl))
                    .cookie(replaceSessionTokenCookie(Optional.of(user)))
                    .build()
            }
        }
        throw WebApplicationException(BAD_REQUEST)
    }

    /**
     * Handles the OAuth server response.
     *
     * @return The OAuth identifier for this user if verification was successful
     */
    @GET
    @Path("/verify")
    fun verifyOAuthServerResponse(@Context request: HttpServletRequest): Response {
        try {
            val user =
                INSTANCE.getBySessionToken(
                    request.cookies
                        ?.firstOrNull { it.name == JavabotConfiguration.SESSION_TOKEN_NAME }
                        ?.value
                )
            if (user != null) {
                user.authorities.add(ROLE_PUBLIC)
                val admin = adminDao.getAdminByEmailAddress(user.email)
                if (admin != null) {
                    user.authorities.add(ROLE_ADMIN)
                } else {
                    if (adminDao.count() == 0L) {
                        adminDao.save(Admin(user.email))
                        user.authorities.add(ROLE_ADMIN)
                    }
                }
                INSTANCE.put(user)
                return Response.temporaryRedirect(URI("/"))
                    .cookie(replaceSessionTokenCookie(Optional.of(user)))
                    .build()
            }
            throw WebApplicationException(UNAUTHORIZED)
        } catch (e: Exception) {
            log.error("OAuth verification failed: {}", e.message, e)
            throw WebApplicationException(UNAUTHORIZED)
        }
    }

    /** @return Get an initialized User from session cookie */
    protected fun replaceSessionTokenCookie(user: Optional<User>): NewCookie {
        if (user.isPresent) {
            val value = user.get().sessionToken.toString()
            log.debug("Replacing session token with {}", value)
            return NewCookie(
                JavabotConfiguration.SESSION_TOKEN_NAME,
                value,
                "/",
                null,
                null,
                86400 * 30,
                false,
            )
        } else {
            // Remove the session token cookie
            log.debug("Removing session token")
            return NewCookie(
                JavabotConfiguration.SESSION_TOKEN_NAME,
                null,
                null,
                null,
                null,
                0,
                false,
            )
        }
    }

    companion object {

        private val log = LoggerFactory.getLogger(PublicOAuthResource::class.java)
        val AUTH_MANAGER: String = "authManager"
    }
}
